src/NuGet.Clients/NuGet.PackageManagement.VisualStudio/Services/NuGetLockService.cs

```diff
@@ -44,42 +44,47 @@ public NuGetLockService(JoinableTaskContext joinableTaskContext)
         /// </summary>
         public async Task<T> ExecuteNuGetOperationAsync<T>(Func<Task<T>> action, CancellationToken token)
         {
-            return await NuGetUIThreadHelper.JoinableTaskFactory.RunAsync(async () =>
+            if (_lockCount.Value == 0)
             {
-                if (_lockCount.Value == 0)
+                return await _joinableTaskFactory.RunAsync(async delegate
                 {
-                    return await _joinableTaskFactory.RunAsync(async delegate
+                    using (_joinableTaskCollection.Join())
```

Comment:
This pattern to join the collection containing the task itself may trigger a performance problem, when it is called a lot.  I think it will be another issue to look at in the future.